sqrl-planner/src/main/java/com/datasqrl/plan/rules/SQRLLogicalPlanRewriter.java

```diff
@@ -267,14 +286,33 @@ public RelNode visit(LogicalFilter logicalFilter) {
     List<TimePredicate> timeFunctions = new ArrayList<>();
     List<RexNode> conjunctions = rexUtil.getConjunctions(condition);
 
-    //Identify any pk columns that are constrained by an equality constrained with a constant and remove from pk list
+
+    //Identify any columns that are constrained to a constant value and a) remove as pk if they are or b) update pk if filter is on row_number
     Set<PrimaryKeyMap.ColumnSet> pksToRemove = new HashSet<>();
+    List<Integer> newPk = null;
     for (RexNode node : conjunctions) {
       Optional<Integer> idxOpt = CalciteUtil.isEqualToConstant(node);
       idxOpt.flatMap(input.primaryKey::getForIndex).ifPresent(pksToRemove::add);
+      if (idxOpt.isPresent() && input.relNode instanceof LogicalProject) {
+        RexNode column = ((LogicalProject) input.relNode).getProjects().get(idxOpt.orElseThrow());
+        if (column instanceof RexOver && column.isA(SqlKind.ROW_NUMBER)) {
+          newPk = ((RexOver)column).getWindow().partitionKeys.stream().map(n ->  CalciteUtil.getInputRefThroughTransform(n, List.of(CAST_TRANSFORM, COALESCE_TRANSFORM)))
+              .map(opt -> opt.orElse(-1)).collect(Collectors.toUnmodifiableList());
+          if (newPk.stream().anyMatch(idx -> idx<0)) {
+            newPk = null; // Not all partition RexNodes are input refs
+          }
+        }
+      }
```

Comment:
This block is a good candidate for an optional stream since theres a lot of boxing and unboxing

Reply:
I see what you are saying, but that variable assignment ruins the stream